Issue 204: Allow max_delay to be more flexible

adrian-lison opened the pull request:
This PR aims to solve #204 by increasing the robustness of `max_delay` specification handling. This is achieved by:

1. Adding more metadata to the preprocessed observation object that distinguish between the maximum delay specified by the user (`spec`), the maximum delay observed in the data (`obs`),  and the maximum delay used in the model (`model`).
2. Adding a default behavior in situations when the specified maximum delay is longer than the observed one. In this case, the observed maximum delay is now used in the model to avoid identifiability problems. In other words, the behavior is now: `model = min(spec, obs)`.
3. Adding checks of the maximum delay with warnings that draw attention to i) a longer specified maximum delay than observed (the warning then suggests adding unobserved delays using `enw_complete_dates`), or ii) a potentially too short specified maximum delay (crude heuristic currently used: warn if more than 50% of reference dates have a maximum `cum_prop_reported` below 80%).
4. Adjusting the behavior of postprocessing functions. Originally, the outputs of `enw_nowcast_summary` and `enw_nowcast_samples` would only go back to the maximum delay used in the `model`. They are now adjusted to go back by the maximum delay specified by the _user_. This makes it easier to join/compare fits from different models with the same _specified_ but a different _observed_ maximum delay. The posterior sampels or summaries for not-modeled delays that need to be added if `spec`>`obs` are as if they came from a fixed variable (all samples identical, hence all measures of location identical, and measures of variation are 0). 

The PR also further improves the function naming and documentation around methods for handling the delays.

ToDos:
- [x] Resolve conflicts
- [x] Update NEWS.md

seabbs commented:
This looks really great @adrian-lison - just going through commit by commit. Just to note that there may be a significant number of merge issues coming your way when #220 gets merged (which I think may be soon). Happy to help resolve those if you would like and if so just give me a ping. 

Does this PR also go some way to addressing #114?

> Adding a default behavior in situations when the specified maximum delay is longer than the observed one. In this case, the observed maximum delay is now used in the model to avoid identifiability problems. In other words, the behavior is now: model = min(spec, obs).

This seems like a good pragmatic choice for the most part and in many model specifications. I think any choice other than this also currently breaks our code? There could be some settings where this is not the desired behaviour though (as relying on the log normal dist for unobserved reports etc) so we could think about making another feature issue for optionally altering this default? The sort of settings I am thinking about would be outbreaks etc where data is initially very spare and we need to rely on priors to some extent.

> Adding checks of the maximum delay with warnings that draw attention to i) a longer specified maximum delay than observed (the warning then suggests adding unobserved delays using enw_complete_dates), or ii) a potentially too short specified maximum delay (crude heuristic currently used: warn if more than 50% of reference dates have a maximum cum_prop_reported below 80%).

These are both really great ideas. Love it. 

> Adjusting the behavior of postprocessing functions. Originally, the outputs of enw_nowcast_summary and enw_nowcast_samples would only go back to the maximum delay used in the model. They are now adjusted to go back by the maximum delay specified by the user. This makes it easier to join/compare fits from different models with the same specified but a different observed maximum delay. The posterior sampels or summaries for not-modeled delays that need to be added if spec>obs are as if they came from a fixed variable (all samples identical, hence all measures of location identical, and measures of variation are 0).

Also a great idea.

adrian-lison commented:
> Does this PR also go some way to addressing #114 (we could just put a note in that issue if so to help whoever ends up making a PR to close it)?

Good point @seabbs, I was also thinking of this while working on the `max_delay` again. We could also try to tackle it within the current PR, but maybe it's better to get this one out first.

adrian-lison commented:
> There could be some settings where this is not the desired behaviour though (as relying on the log normal dist for unobserved reports etc) so we could think about making another feature issue for optionally altering this default? The sort of settings I am thinking about would be outbreaks etc where data is initially very sparse and we need to rely on priors to some extent.

But wouldn't it always be better to just ask the user to padd their data with zeros for unobserved delays? In what case should we specifically account for "unknown" reports beyond an observed maximum delay?

seabbs commented:
> But wouldn't it always be better to just ask the user to padd their data with zeros for unobserved delays? 

Doing this is the right call when they are not observed but we know they are zero. If you do that when they may not be zero you will bias the model as the delays you have imputed will be included in the model. 

An example is if we only have 5 days of data but we expect the delays to be as long as 30 days. Here you could assume say a log normal for the delay with a strong prior and those first 5 days will be well fit and perhaps influence the unobserved remainder. Under your proposal the max delay would 5 and so you would only be able to now a 5 day horizon vs the expected final horizon. Adding zeros would then cause bias as it would assume no reports rather than some unknown reports.

adrian-lison commented:
> An example is if we only have 5 days of data but we expect the delays to be as long as 30 days. Here you could assume say a log normal for the delay with a strong prior and those first 5 days will be well fit and perhaps influence the unobserved remainder. Under your proposal the max delay would 5 and so you would only be able to now a 5 day horizon vs the expected final horizon. Adding zeros would then cause bias as it would assume no reports rather than some unknown reports.

Okay, so you don't mean left-truncation here, but rather sth. like when we are at the start of a new epidemic and the earliest reference date that exists is smaller than the supposed max delay, right?

You said the model currently breaks down if there are unobserved delays - would we need changes in the model or is it just a matter of the prior?

adrian-lison commented:
@seabbs Would you mind checking out the tests, I'm having a bit of a hard time getting all the snapshots up-to-date...

seabbs commented:
> Would you mind checking out the tests, I'm having a bit of a hard time getting all the snapshots up-to-date...

Yes, will do. They are a pain.

seabbs reviewed (changes requested):
Thanks @adrian-lison - nice one. I have a few major concerns that I will list here and then some minor points (in review comments). Major concerns: 

1. Use of a `list` vs a `data.table` for `max_delay`. This breaks package convention and I think makes it slightly harder to communicate clearly what we are doing. Also moving to this means that what the `max_delay` actually is isn't immediately readable from the preprocessing object. I think this again highlights our need for a custom print function here and a general redo of proprocessing and the `epinowcast` object.
2. On a similar theme I think `check_max_delay` doesn't currently make it as clear as we might be that we are resetting `max_delay` as specified by the user. Does it also feel like this function should be called by `enw_metadata_maxdelay` vs on its own?
3. A few functions have been renamed here which is fine and sensible but is still a breaking change. We need to at least carefully broadcast this in the news but ideally would use `lifecycle` depreciation as I have done in the data converters draft work. 
4. I don't really like requiring the user to pass `max_delay` to post-processing functions. Does this feel like it benefits an edge case whilst being quite a large negative all of the rest of the time? Could we rather provide some help linking together model fits with different maximum delays instead as a helper function vs doing this here?

seabbs reviewed (changes requested):
Thanks @adrian-lison - nice one. I have a few major concerns that I will list here and then some minor points (in review comments). Major concerns: 

1. Use of a `list` vs a `data.table` for `max_delay`. This breaks package convention and I think makes it slightly harder to communicate clearly what we are doing. Also moving to this means that what the `max_delay` actually is isn't immediately readable from the preprocessing object. I think this again highlights our need for a custom print function here and a general redo of proprocessing and the `epinowcast` object.
2. On a similar theme I think `check_max_delay` doesn't currently make it as clear as we might be that we are resetting `max_delay` as specified by the user. Does it also feel like this function should be called by `enw_metadata_maxdelay` vs on its own?
3. A few functions have been renamed here which is fine and sensible but is still a breaking change. We need to at least carefully broadcast this in the news but ideally would use `lifecycle` depreciation as I have done in the data converters draft work. 
4. I don't really like requiring the user to pass `max_delay` to post-processing functions. Does this feel like it benefits an edge case whilst being quite a large negative all of the rest of the time? Could we rather provide some help linking together model fits with different maximum delays instead as a helper function vs doing this here?

seabbs commented:
Hi @adrian-lison,

I have some time - would it be helpful for me to cran through some of the above comments and make suggested changes?

seabbs commented:
I will circle back and resolve these soon

adrian-lison commented on R/postprocess.R:
```diff
@@ -83,8 +86,8 @@ enw_posterior <- function(fit, variables = NULL,
-#' enw_nowcast_summary(fit$fit[[1]], fit$latest[[1]])
-enw_nowcast_summary <- function(fit, obs,
+#' enw_nowcast_summary(fit$fit[[1]], fit$latest[[1]], fit$max_delay[[1]])
+enw_nowcast_summary <- function(fit, obs, max_delay,
```
Yeah, I didn't consider that this method would also be directly called by the user. The reason why it is there is that we need this to add artificial summary statistics for not-modeled dates as described further above. The goal is to make it easy to join/compare fits from different models with the same specified but a different observed maximum delay, which could occur frequently.

Would it be an alternative to pass the whole `pobs` object all the time?

seabbs replied:
> The reason why it is there is that we need this to add artificial summary statistics for not-modeled dates as described further above. The goal is to make it easy to join/compare fits from different models with the same specified but a different observed maximum delay, which could occur frequently.

Yeah, I get that and it makes sense but is this the most common issue people have? Could we instead provide a function to help join (and make everything cover a consistent period?). We could of course require people to pass the whole  `pobs` object but the idea of the lower level functions is that people have more control of the inputs. If we go with this need to at least make it optional I think?

adrian-lison replied:
Okay, so two options

1. Make max_delay optional, only when this is specified the non-modeled delays are added
2. Write a separate function called something along the lines of "set max_delay" that you can apply to the output of enw_nowcast_summary

I think I would be happy with option 1, what do you think?

adrian-lison replied:
@seabbs :)

adrian-lison commented on R/postprocess.R:
```diff
@@ -94,15 +97,30 @@ enw_nowcast_summary <- function(fit, obs,
+  nowcast <- rbind(obs_head, nowcast, fill = TRUE)
+  
+  # add artificial summary statistics for not-modeled dates
+  nowcast[1:nrow(obs_head), c("mean", "median") := confirm]
```
Spurious in what sense? This is assuming that non-modeled delays are beyond the maximum delay, so it is natural to predict exactly the observed values?

While writing this I am starting to worry about the missing reference date case...

seabbs replied:
Okay, I have no idea why I thought this. Seems fine looking at it now.

adrian-lison replied:
But I will have to think about the missing reference dates. If there are any, then we would still want an uncertain nowcast there. How do we currently handle this with the missingness model again, i.e. how far backwards do we model the missing reference dates and report them with the nowcast?

seabbs replied:
Sorry missed this round of questions. We only model missing references dates from the start of observed reference dates and we only include them in the likelihood when they have been completely observed (i.e up to the max delay).

adrian-lison commented on R/preprocess.R:
```diff
@@ -810,8 +807,8 @@ enw_missing_reference <- function(obs) {
-#' enw_delay_metadata(20, breaks = 4)
-enw_delay_metadata <- function(max_delay = 20, breaks = 4) {
+#' enw_metadata_delay(max_delay = 20, breaks = 4)
+enw_metadata_delay <- function(max_delay = 20, breaks = 4) {
```
So, do you think changing the naming is worth it? And how risky is it, are users actually using these function directly? They seemed very internal to me...

seabbs replied:
yeah I like it but as suggested think we should use `lifecycle`. Also chatted a bit with Carl about what we should do with internal/external. Before we can decide that I think we need to make some progress on any redesign of preprocessing + try and understand some use in the wild.

adrian-lison commented on R/preprocess.R:
```diff
@@ -892,7 +925,7 @@ enw_construct_data <- function(obs, new_confirm, latest, missing_reference,
     by = list(by),
     groups = length(unique(obs$.group)),
-    max_delay = max_delay,
+    max_delay = list(max_delay),
```
This could make sense.

adrian-lison commented on R/check.R:
```diff
@@ -151,15 +156,60 @@ check_modules_compatible <- function(modules) {
+#' Compare maximum delays specified by the user vs. observed in the data, and
+#' raise potential warnings.
+#'
+#' @param latest_obs The latest available observations.
```
True, it only works after `enw_add_max_reported` has been called on the `obs`.

seabbs replied:
cool makes sense. Can also use `coerce_dt()` to do some input checking if feeling fancy and if feeling like its user exposed enough to need it.

adrian-lison commented on R/preprocess.R:
```diff
@@ -983,8 +1019,12 @@ enw_preprocess_data <- function(obs, by = c(), max_delay = 20,
   obs <- enw_add_delay(obs)
+  
+  # add metadata about different maximum delays
+  max_delay <- enw_metadata_maxdelay(obs = obs, max_delay = max_delay)
```
Further argument for calling it `metamaxdelay`

seabbs commented:
> I see, we can just change this to a data.table then I guess.

I agree its not ideal but probably a better choice for now? We need to rethink the whole object I think (or at least how it is presented to the user).

seabbs commented:
Would it be helpful if I handled rebasing this PR onto our current `main`?

seabbs commented:
I rebased this one main and did some linting

adrian-lison commented on R/preprocess.R:
```diff
@@ -950,13 +985,14 @@ enw_construct_data <- function(obs, new_confirm, latest, missing_reference,
+#' nowcasting.
+#' - `groups`: Numeric, Number of groups/strata in the supplied observations 
+#' (set using `by`).
+#' - `max_delay`: A `list`, with the (potentially different) maximum delays 
```
true...

seabbs commented:
I'd really like to wrap `0.2.1` so I can do some other projects for a bit before coming back to dev work. Do you think we can get through this tomorrow (Monday)? I am happy to do some/all the leg work but we will probably need to chat here about features/choices etc. 

Not a problem if not as we can just push into the next release and then there is no time pressure.

adrian-lison commented:
I think we should better have this for the next release. Most things are easy, but one or two features still need a bit of careful design

seabbs commented:
> I think we should better have this for the next release. Most things are easy, but one or two features still need a bit of careful design

Yes, agree. I looked it over and definitely still need to think about a few things vs it just being implementation.

seabbs commented:
@adrian-lison anything I can do here to help unstick this?

adrian-lison commented:
Note to myself, ToDos for finishing this PR

- [ ] Add lifecycle for `enw_filter_delay` and `enw_metadata_delay`
- [ ] Change `enw_metadata_maxdelay` return object to `data.table`, and non-abbreviated names, and long-formatted description variable with explanation
- [ ] Postprocessing of nowcast_samples and nowcast_summary: Make `max_delay` optional, only when this is specified the non-modeled delays are added
- [ ] Make new issue for handling of postprocessing under missing reference dates. If there are any, then we would still want an uncertain nowcast there, even if the modeled maximum delay was shorter?
- [ ] Generalise check_max_delay to be user-facing and have an output to help here (i.e what is the cumulative density of my delay by group?).
- [ ] Rename `max_delay` to `metamaxdelay` in preprocessing object and refactor where it is used.
- [ ] Clarify the warning in `check_max_delay` which is saying that the max_delay specified by the user is ignored and we instead use the maximum observed one.
- [ ] Call `check_max_delay` inside `enw_metadata_maxdelay`

seabbs commented:
Anything I can do here to help unstick this?

adrian-lison commented:
A few side remarks:
- [Made enw_add_max_reported robust to multiple calls](https://github.com/epinowcast/epinowcast/pull/224/commits/fcbda38a2259559c8317fb0ab5750d3f98954789)
- [Suppressed deprecation message from rstan::extract_sparse_parts](https://github.com/epinowcast/epinowcast/pull/224/commits/5aea993ca5d8dd92898564ebca3a2ff4dde3ca64). I wrote a PR for `rstan` that fixes the deprecation issue: https://github.com/stan-dev/rstan/pull/1083

adrian-lison commented:
I get this weird warning on R-CMD-check

```
Execution halted
  enw_report: no visible binding for global variable ‘type’
    (/Users/runner/work/epinowcast/epinowcast/check/epinowcast.Rcheck/00_pkg_src/epinowcast/R/model-modules.R:187-188)
  Undefined global functions or variables:
    type
```

I don't get this, `type` is only used as a `data.table` column in this function...

seabbs commented:
> I get this weird warning on R-CMD-check

This a weird quirk of R CMD check where it thinks things inside `data.table` calls or in `dplyr` calls are global variables. The "solution" is to add `type` as a global variable in `utils.R` or define it as null before it is used in the function that is a problem. Pretty silly stuff.

adrian-lison commented:
@seabbs Calling check_max_delay inside enw_metadata_maxdelay is unfortunately not really possible as it requires the `latest` dat.table, which is only constructed later in the preprocessing function and depends on `metamaxdelay`

adrian-lison commented:
@seabbs I think this is ready for another review. :) I ended up changing `check_max_delay` quite a bit while trying to make it user-facing.
- It now expects the normal `obs` and a `max_delay` as input, which makes it more generic to use. For example, a user can try out different `max_delay` values on the `obs` to see what the covered proportions are.
- I first implemented an output that gives the `low_cum` values by group, but ended up reverting that. It now only gives the total `low_cum` value, which avoids issues with different groupings before and after `add_incidence` is called in the preprocessing script
- With the above changes, `check_max_delay` can now indeed be called in `enw_meta_maxdelay`, and I use it to add `low_cum` values for each maximum delay (spec, obs, model) to the metadata `data.table`

One thing to discuss maybe is that the proportion of reference dates for which the maximum delay is too short is currently computed over all reference dates in `check_max_delay`. However, as this uses the `cum_proportion` values, which are based on observed maximum counts, reference dates that are less than `max_delay` in the past will always have complete coverage by definition. This biases the `low_cum` value downwards. What we could do alternatively is to only use reference dates which are at least `max_delay` days before the latest observed report date in `obs`. Obviously, this will only work if the `obs` are long enough... But otherwise checking the maximum delay doesn't make sense anyways and could give a false sense of safety?

adrian-lison commented on R/check.R:
```diff
@@ -205,3 +205,69 @@ coerce_dt <- function(
+    obs, max_delay = max_delay
+  )
+  
+  # filter by earliest observed report date
+  obs <- obs[,
+     .SD[reference_date >= min(report_date) | is.na(reference_date)],
+     by = .group
+  ]
```
I adopted this from `enw_add_incidence`. However, I didn't quite like it there, because it seemed like an unexpected side effect that is not directly related to adding incidence/diff_obs to the obs. So I was wondering if we could maybe move this out of `enw_add_incidence`, perhaps into a separate function?

adrian-lison replied:
Made an issue for it in #305

seabbs commented:
> I first implemented an output that gives the low_cum values by group, but ended up reverting that. It now only gives the total low_cum value, which avoids issues with different groupings before and after add_incidence is called in the preprocessing script

Could we make this optional? It sounds like quite a nice feature.

seabbs commented:
> Added the internal keyword to all deprecated functions so they don't appear in the documentation index anymore, but the documentation can still be accessed for the functions directly

ooo nice!

seabbs commented:
> file=R/data-converters.R,line=231,col=24,[implicit_assignment_linter] Avoid implicit assignments in function calls. For example, instead of `if (x <- 1L) { ... }`, write `x <- 1L; if (x) { ... }`.

This one looks like we can just avoid using  # nolint as for our actual use case its fine

seabbs commented on R/model-modules.R:
```diff
@@ -439,16 +441,16 @@ enw_expectation <- function(r = ~ 0 + (1 | day:.group), generation_time = 1,
     !(as_string_formula(formula) %in% "~0")) {
-    stop("A missingness model has been specified but data on the proportion of
-          observations without reference dates is not available.")
+    stop(paste("A missingness model has been specified, but no observations", 
```
no paste needed.

seabbs commented on R/postprocess.R:
```diff
@@ -92,15 +95,47 @@ enw_nowcast_summary <- function(fit, obs,
 
   ord_obs <- coerce_dt(obs)
-  ord_obs <- ord_obs[reference_date > (max(reference_date) - max_delay)]
+  ord_obs <- ord_obs[reference_date >
```
for readability can we link break this from just before reference date?

seabbs commented on R/postprocess.R:
```diff
@@ -92,15 +95,47 @@ enw_nowcast_summary <- function(fit, obs,
-  )
+  obs_model <- ord_obs[reference_date >
+    (max(reference_date) - max_delay_model)]
+  obs_spec <- ord_obs[reference_date <=
```
same thing here

seabbs commented on R/postprocess.R:
```diff
@@ -124,34 +162,60 @@ enw_nowcast_summary <- function(fit, obs,
     id.vars = c(".chain", ".iteration", ".draw")
   )
-  max_delay <- nrow(nowcast) / (max(obs$.group) * max(nowcast$.draw))
+
```
as this is duplicated across functions I think it could do with being an internal function to reduce technical debt.

seabbs replied:
We also see a coverage decrease due to the duplicated code for checking the meta delay. Perhaps if we make that an internal as well and test it we can kill two birds with one stone?

adrian-lison replied:
Maybe rather not, because the two lines are not identical, in the case of enw_nowcast_samples we also need to divide by the number of  draws, I don't really see why writing a function for two different one-liners that are quite verbatim

seabbs commented on R/preprocess.R:
```diff
@@ -329,16 +332,21 @@ enw_add_delay <- function(obs, copy = TRUE) {
     ,
     .(reference_date, .group, max_confirm = confirm)
   ]
+  if ("max_confirm" %in% colnames(obs)) {
```
can we use coerce_dt for this?

seabbs commented on R/preprocess.R:
```diff
@@ -480,10 +488,9 @@ enw_latest_data <- function(obs) {
 
   latest_data <- latest_data[,
-    .SD[report_date == (max(report_date)) | is.na(reference_date)],
+    .SD[report_date == (max(report_date)) & !is.na(reference_date)],
```
well this is all very strange! Why not filter out missing reference dates from the get go?

seabbs reviewed (commented):
This is looking really good I think. Just a few style bits and bobs to cover I think. 

I've also only sight read so far and would like to play through using the PR - will do that tomorrow.

seabbs commented on R/check.R:
```diff
@@ -182,26 +182,96 @@ coerce_dt <- function(
+  # warnings obtained would also apply to a modelled, potentially shorter,
+  # maximum delay.
+  obs <- enw_filter_delay(
+    obs,
```
I reverted but its just personal preference. open to changing back

seabbs commented:
Coverage decrease is just from the deprecated functions I think so not a worry.

seabbs commented:
went through and fixed the remaining linting issues. 🤞🏻 we will have some passing checks now!

seabbs commented:
So I think we have a problem that is being flagged up by the checks. 

In the example for `check_max_delay()` the input has strata but no grouping variable and this causes the function to fail. We could do a few things to fix this. 

Firstly as discussed above we could allow `check_max_delay()` to have a by-argument and we could check for duplicate reference and report date pairs as you have added elsewhere and error if these exist.

We could also just check for duplicates and a group variable and error if this condition isn't met. 

Or we could put a lot of documentation signposting in...

I think probably the first one is the better option (though it takes more work sadly) - this would also make it easier for another PR to make the output groupwise which does seem like it would be handy/

adrian-lison commented:
> > I first implemented an output that gives the low_cum values by group, but ended up reverting that. It now only gives the total low_cum value, which avoids issues with different groupings before and after add_incidence is called in the preprocessing script
> 
> Could we make this optional? It sounds like quite a nice feature.

Implemented.

> > What we could do alternatively is to only use reference dates which are at least max_delay days before the latest observed report date in obs.
> 
> I think this is the best option. We could then throw a warning about there not being enough data to compute this?

Implemented, but note that I am using the _maximum observed delay_ to restrict the reference dates. Using the `max_delay` input to the function wouldn't have made sense. I implemented some version with quantiles to increase robustness to outliers.

adrian-lison commented:
> So I think we have a problem that is being flagged up by the checks.
> 
> In the example for `check_max_delay()` the input has strata but no grouping variable and this causes the function to fail. We could do a few things to fix this.
> 
> Firstly as discussed above we could allow `check_max_delay()` to have a by-argument and we could check for duplicate reference and report date pairs as you have added elsewhere and error if these exist.
> 
> We could also just check for duplicates and a group variable and error if this condition isn't met.
> 
> Or we could put a lot of documentation signposting in...
> 
> I think probably the first one is the better option (though it takes more work sadly) - this would also make it easier for another PR to make the output groupwise which does seem like it would be handy/

I agree, and I would say that changing this to support a "by" argument makes sense. We should strive for making this function work with data before it is put into `enw_preprocess`, because in `enw_preprocess` the user has to specify the `max_delay`, so it would be nice to be able to use `check_max_delay` beforehand to guide the decision.

I implemented additional operations to achieve the above aim now, except for throwing an error when the data is stratified by more variables than mentioned in `by`.  Would add this after #295 is merged. Unfortunately this now requires quite some repetitions from the `enw_preprocess_data` function. @seabbs maybe have a look and let me know if we should try to outsource these first parts of `enw_preprocess_data` to be able to reuse them (we would then want to make this function have an optional maximum delay).

seabbs commented:
Got a bit overwhelmed and scared of this PR but will circle back soon. Having thoughts that me making you make the check function standalone was a very very bad call on my part...

adrian-lison commented on R/model-module-helpers.R:
```diff
@@ -217,7 +217,7 @@ add_pmfs <- function(pmfs) {
 #' extract_sparse_matrix(mat)
 extract_sparse_matrix <- function(mat, prefix = "") {
-  sparse_mat <- rstan::extract_sparse_parts(mat)
+  sparse_mat <- suppressMessages(rstan::extract_sparse_parts(mat))
```
@seabbs We could do this, but if we copied over code from rstan we would have a license issue since they use GPL, and we use MIT currently. So we should either code something similar on our own or adapt our license, which would make the package more restrictive, however.

adrian-lison commented:
Based on a discussion with @seabbs, the plan is now to remove the automatic choice of the observed maximum delay but keep the internals of informing about the differences between delays. I have come up with the following todos:

ToDos

- [ ] - Revert `check_max_delay` to be internal again and work with the preprocessed object (apply at later point in `enw_preprocess_data`)
- [ ] - Make `check_max_delay` error messages less cryptic (considering that the function will now likely not be called directly by the user)
- [ ] - Change `enw_metadata_maxdelay` to always model the user-specified max delay (but provide a warning, and hint at metamaxdelay)
- [ ] - Remove "specified" type from `metamaxdelay` because this will always be identical to the modeled one in the new version
- [ ] - Adjust postprocessing functions to take as argument simply a `max_delay`,  not a `metamaxdelay` object. This allows users to more easily supply a custom max delay when comparing different models. By default, the modelled maximum delay is passed.
- [ ] - Adjust `enw_preprocess_data` accordingly

How to help users change the maximum delay to the observed one?
- Option 1: Add helper function that can be applied to the preprocessed object to update the max delay (it takes pobs, and returns a new pobs). This would be very simple because it can use `pobs$metamaxdelday` to get the observed max delay, and then just call `enw_preprocess_data` a second time with `pobs$obs` as data input?
- Option 2: Add an option to automatically chose the observed maximum delay, for example by providing `max_delay="observed"` instead of a number to the `enw_preprocess_data` function. Or with a dedicated helper function.

Feedback welcome!

seabbs commented:
>  Add helper function that can be applied to the preprocessed object to update the max delay (it takes pobs, and returns a new pobs). This would be very

It feels like it would be very clean to add a method for `update` for the preprocessed data class. Then you would just call `update` on `pobs` with your new delay (or grouping , data etc.) Not sure we need to make that auto discoverable from the current `pobs` but maybe?

adrian-lison commented:
> > Add helper function that can be applied to the preprocessed object to update the max delay (it takes pobs, and returns a new pobs). This would be very
> 
> It feels like it would be very clean to add a method for `update` for the preprocessed data class. Then you would just call `update` on `pobs` with your new delay (or grouping , data etc.) Not sure we need to make that auto discoverable from the current `pobs` but maybe?

I can have a look at this. It just requires us to have all inputs to the preprocessing function stored in unaltered form in pobs. Which is desirable anyway?

adrian-lison commented:
@seabbs Finally found time to work on this. I actually ended up removing the `metamaxdelay` object altogether, since it didn't seem useful anymore after removing the distinction between "specified" and "modelled", and just made the code more complex. Also, this now means that all checks related to the max delay are really in `check_max_delay`.

The latter is now again a function that takes the preprocessed obs as input. I currently call it at the end of `enw_preprocess` now (on `out[]`). I thought this is convenient, we can add also more check functions to the end of `enw_preprocess`, and they may all use pobs as input. At the same time, they can also be called by the user by passing a pobs object.

Now what we still have to decide on is how to provide a helper to change the maximum delay to the observed one etc. I generally find your idea of an update function for pobs interesting, but propose not to tackle this within this PR but make a new PR for it. So I would rather write a standalone helper just for the max delay now and we can integrate this into a generic update function as part of another PR.

Finally, what do you think about offering `max_delay="observed"` as a valid input to `enw_preprocess` that then automatically chooses the maximum observed delay?

adrian-lison commented:
Remaining ToDos

- [ ] ensure that doc and messages in `check_maxdelay` make sense now (behaviour when specified delay is short or longer than observed)
- [ ] Add `max_delay="observed"` option to `enw_preprocess`
- [ ] Implement helper to update `max_delay` in pobs
- [ ] Make PR for generic update of pobs
- [ ] merge main into this branch, ensure that especially #351 and #346 integrate correctly
- [ ] update NEWS

adrian-lison commented:
@seabbs Thinking again about this, I would rather not work on any updating functionality for `pobs` in this PR, since it was not the original aim and I'd love to merge this finally.

adrian-lison commented:
I merged in the current main branch and resolved all conflicts. The new `timestep` feature gave me a bit of a headache but now everything should work properly.

adrian-lison commented:
Note that I removed the warnings from `enw_filter_delay` for empirical max delay being smaller than the specified max delay, because this should now be handled by `check_max_delay`.

adrian-lison commented:
@seabbs I implemented the `max_delay = "observed"` option and made it the new default in `enw_preprocess_data`. This seems more robust to me than the old default of 20 days. This is obviously a breaking change, but there is now a warning message when using `max_delay = "observed"` anyway, so users should become aware.

I hope the warnings are informative now. I noticed that we have some inconsistencies in the documentation with regards to max_delay and timestep, because when a non-daily timestep is used, the max_delay is not in days any more. Code-wise, everything should be consistent, but the docs need some work. I will make a separate issue for this though.

adrian-lison commented:
@seabbs What do you think?

seabbs commented:
Sorry about the delay on this. I just fixed the merge issues and will circle back later today.

adrian-lison commented:
@seabbs Do you need any kind of help?

adrian-lison commented:
Hmm, a bit unsure what the problem is. The failing touchstone / Benchmarks checks are fixed. But now some checks are failing in the tests because some figures do not match (Doppelganger). But this seems to depend on the OS version?

seabbs commented:
This was due to the new version of `ggplot2`. I fixed this in #443 and just resolved the merge issue you had here so hopefully everything should now pass.

adrian-lison commented:
@seabbs I just updated from main, now for some reason it's asking again for you to approve changes, otherwise it won't try to run the checks...

seabbs commented:
I just triggered them manually for fingers crossed - this is due to the README auto update (i.e #447)

seabbs commented on R/check.R:
```diff
@@ -222,30 +224,227 @@ coerce_dt <- function(
+  }
+  timestep <- data$timestep
+
+  if (warn == "internal") {
```
Do we really need this complexity here vs just setting `warn = FALSE` when used internally? Generally I am a bit uncomfortable with arguments that change type (i.e here from logical to character)

adrian-lison replied:
I feel like the case distinction is useful as we also want warnings when used internally, but slightly different warning messages compared to direct use of the function. I addressed the type change issue by adding a separate argument warn_internal that is FALSE by default and explaining it in the doc. Should be cleaner now.

seabbs commented on R/check.R:
```diff
@@ -222,30 +224,227 @@ coerce_dt <- function(
+                            warn = TRUE) {
+
+  if (is.null(max_delay)) {
+    max_delay <- data$max_delay
```
Rather than having this here I think you can set it as the default above (i.e max_delay= data$max_delay)

adrian-lison replied:
True, I keep forgetting this is possible

seabbs commented on R/check.R:
```diff
@@ -222,30 +224,227 @@ coerce_dt <- function(
+  if (!(maxdelay_quantile_outlier > 0 && maxdelay_quantile_outlier <= 1)) {
+    cli::cli_abort(
+      "`maxdelay_quantile_outlier` must be between 0 and 1, e.g. 0.97 for 97%."
+      )
```
```suggestion
    )
```

seabbs commented on README.md:
```diff
@@ -194,6 +194,8 @@ list.files(
 
 </details>
 
+## Contributors
```
hmm why is this showing up as a new change?

adrian-lison replied:
Good catch, looks like the merging failed and it kept the old Contributors label (now it's called Contributing). I removed the old one now.

seabbs reviewed (commented):
This is looking really good now (sorry it has been such an incredibly long mission). I have a last few comments about input arg choices but functionality wise I am very happy. Will dig in to CI and checks etc tomorrow and may suggest some code changes related to that if anything comes up

adrian-lison commented on R/model-modules.R:
```diff
@@ -551,7 +553,7 @@ enw_missing <- function(formula = ~1, data) {
     # Get report dates that cover all reference dates up to the max delay
     rep_w_complete_ref <- enw_reps_with_complete_refs(
       data$new_confirm[[1]],
-      max_delay = data$max_delay[[1]],
```
Max delay is not a list anymore, so I don't think we need the indexing? It is now just
`max_delay = data$max_delay`
Or do you think this is not safe?

seabbs replied:
I was thinking because the input is always a data.table of row one and therefore I thought you had to index it but you are right this is working and it doesn't seem unsafe.

seabbs reviewed (dismissed):
This is looking good now I think! @adrianlison it would be good if you could give this a careful check given its a lot and the following changes,

- Made the default in `enw_preprocess_data(max_delay..)` to be missing vs using "observed" to avoid the change of input arg type. 
- Updated all snapshot tests - these all seem due to the change in default for `max_delay`
- Fixed `summary.epinowcast` where the use of `with` was causing problems. 
- Caught a stray bracket.

I've checked functionality both in the getting started example and in the delay is longer than the data example.

seabbs commented:
:party

adrian-lison reviewed (commented):
I checked all the new changes, everything looks good to me now. Thanks for the last improvements / fixes. Really excited to merge this!! 

 I can't approve as the author of the PR. @seabbs Can you self-review or overwrite the merge block?

seabbs commented:
I just tweaked branch protection this once to override. Woo!

seabbs commented:
What an epic that turned out to be!
